### ISSUE TITLE
NixOS: warn if new release is out

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nixpkgs#20182]
I propose we add some functionality to warn on new NixOS stable release. No idea how to implement that sanely atm.

**Comment by vcunat** [comment]
Hmm, perhaps, both for the big bumps once in several months and for regular channel updates.

The main question is IMO **how** to present the warning to the user(s).

**Comment by NeQuissimus** [comment]
Ubuntu and Amazon Linux present a message when you SSH in.
Oh-My-Zsh executes a check every time you open a shell.

The truly challenging part would be presenting a notification to UI-only users...

A simple config option would be enough to turn this behaviour off.

**Comment by RonnyPfannschmidt** [comment]
im not aware of a ui-only way to manage packages/updates on nix,
a simple/easy way to make this work for the normal channels would be to include a file/property denoting the current stable release, then nixpkgs could put it into a motd as soon as it doesn't match the current system

**Comment by danbst** [comment]
IMO this should be part of `nix-channel` or `nixos-rebuild --upgrade`, and not as login message. If you use `nix-channel --update` to maintain up-to-date system, then you would like to know whether your current nixpkgs is outdated.

If you don't do that, there are lots of chances you don't care about up-to-dateness of yours OS (and don't care about new nixos)

**Comment by vcunat** [comment]
Hmm, yes, general `nix-channel --check` might be nice, especially if avoiding download of the whole tarballs.

**Comment by Profpatsch** [comment]
(triage) has this been implemented?

**Comment by tomberek** [comment]
(triage) close as old.

**Comment by danbst** [comment]
No, don't close. How I envision this to be implemented:

1. Add a script `is-nixos-up-to-date`. It should be able to detect a) which channels are checked out; b) which channel is NixOS built from; c) how old (in time and commits) are current channels behind; d) which NixOS release are maintained and how long
2. when `system.nixos.notifications.enable = true;`, run that script on some interval. Maybe schedule extra run after each `system.autoUpgrade`. Saved output in JSON somewhere in `/var/nixos`
3. When `system.nixos.notifications.loginShell.*` are set, add appropriate notifications to login shell init script. When `system.nixos.notifications.X11.*` - it will add notifications via libnotify, if X is available.
4. Possible options for notifications are: `new-release`, `is-supported`, `is-old` (to show something for unstable channels), `weekly-news`, `reboot-required` (or `outdated-services`, for services which were not restarted during system activation. For example, it is newbie error to assume that update of kernel on Linux is done without reboot - on NixOS you still have to reboot; display manager is not restarted on `nixos-rebuild switch`)
5. For users of `nix-channel`: run that script in the end of `nix-channel --upgrade` and show only `new-release` and `is-supported` notices.

Also, not sure if this has to be per-user.

**Comment by edolstra** [comment]
This has some overlap with https://github.com/NixOS/nixpkgs/pull/23590.

**Comment by florianjacob** [comment]
still important to me

**Comment by florianjacob** [comment]
still important to me. One of my machines is still on 20.03 as I couldn't immediately update due to breakage, and I might forget that one again.

**Comment by RaitoBezarius** [comment]
Still important.

**Comment by Atemu** [comment]
I don't really like the idea of fetching the information about which release is the newest as that requires an online service, a connection to be made to it and of course all of that needs to be implemented. Seems like overengineering to me.

I think we should simply put an eval warning into a NixOS channel when it goes EOL. On every rebuild using such an EOL channel, the user would then see the warning where they'd be directed to https://nixos.org where they could inform themself about what is newest release, what's new in it, what the breaking changes are etc. and then manually update.

**Comment by Atemu** [comment]
We could even go as far as aborting eval and requiring a specific option to be set to allow the build despite the EOL channel but I don't think that's necessary.

**Comment by RonnyPfannschmidt** [comment]
@Atemu your message is confusing, the suggestion was to make the current stable part of the nix expressions served by the channels so a rebuild system could indicate the need for updates, not a network round-trip

**Comment by Atemu** [comment]
@RonnyPfannschmidt ah, I didn't see your comment. A MOTD also isn't bad but I think I like the eval warning/error better.

**Comment by LunNova** [comment]
Some specific ideas:

* final commit marking it past support after the branch is EoL which adds a trace/warning during evaluation, with optout mechanism
* warning in nixos activation script if date is too far in the future (check against SUPPORT_END= date in `/etc/release`), with optout mechanism

**Comment by vcunat** [comment]
I don't think we need any complicated mechanisms.  If you're using a version that's too far in the past, it's a red flag by itself.  Even if simple `nixos-rebuild --upgrade switch` would fix it.  I think the hardest part is how to show such warnings.

**Comment by Atemu** [comment]
@vcunat the problem is how to make users aware that they're using an old version.

`nixos-rebuild --upgrade` also won't upgrade users to new releases. The user has to find out that there even is a new release and then find the steps to upgrade to it all on their own.  
To us involved upstream, those two are trivial matters but that's not the case for every NixOS user, likely not even most of them at this point; certainly not when they're new to this. We should at the very least leave some breadcrumbs.

I think an eval warning that is pushed onto EOL channels which informs the user of the currently eval'd release's EOL status and how to upgrade would be an easy method to achieve this.  
It's hard to miss a big red warning IMHO.

**Comment by vcunat** [comment]
I meant that we can warn if you're running a version tagged too long in the past.

**Comment by Atemu** [comment]
Just today someone found out that they've been using 23.11 for almost the entire 24.05 release cycle:

https://fosstodon.org/@adingbatponder/113460535537017198

In that thread I propose to utilise desktop notifications to notify users of the new release. I propose to limit that to sessions where we can expect to have a notification daemon (i.e. GNOME, KDE) though.

Thinking about the implementation, this notification should pop up on every session login. It should inform the user that their release is out of date and urge them to update in order to stay secure. It should then point them at the release notes which in turn tell them how to update to a newer release and of course all the breaking changes and whatnot. It'd be great if clicking the notification could automatically open the release announcement in a browser.  
It should also tell the user how to turn the warning off because there are legitimate use-cases for running an old NixOS without being nagged to update all the time eventhough we of course don't support that.

What we could also look into for server systems would be to add an SSH MOTD with the same message via `services.openssh.banner`.

Users who DIY their desktop won't be notified here, so it's not a perfect solution but I also think they're the least likely to need this sort of nudge.

This module would be enabled on the stable channel when the next release is officially out by the outgoing release manager as a sort of last act.
